Remove ambiguity from "not safe to write" compiler error message (#4299)
ponylang/ponyc: .release-notes/4299.md

> ## Fix vague error message when it is not safe to write

Reviewer: ```suggestion
## Remove ambiguity from "not safe to write" compiler error message
```

> Added left side type to a "not safe to write" error message.

Reviewer: This is likely to be lacking in meaning for a reader of the release notes who doesn't have your context.

Please include what the error message used to be, why that was vague, and how the new version of the message fixes the vagueness.

## Remove ambiguity from "not safe to write" compiler error message

> Currently, when it is not safe to write, the error message only talks about the right side, not the left side, nor the reason for the failure of safety.

Reviewer: ```suggestion
Previously, when displaying the "not safe to write" error message, the information provided was incomplete in a way that lead to ambiguity. A "not safe to write" error involves two components, the left side type being assigned to and the right side type being assigned to the left side. We were only displaying the right side, not the left side type, thus making it somewhat difficult to know exactly what the error was.
```

As an example, when trying to write to a field in an iso class instance, we get:

```
Error:
/Users/jairocaro-accinoviciana/issues-pony-4290/demo.pony:10:11: not safe to write right side to left side
    x.foo = foo
          ^
    Info:
    /Users/jairocaro-accinoviciana/issues-pony-4290/demo.pony:9:14: right side type: Foo ref
        let foo: Foo ref = Foo
                 ^
```

> 
> 
> With this change, the error becomes:

Reviewer: ```suggestion

The updated error message is now in the format:
```
